[build-script] Add buildbot_linux,lld preset

etcwilde opened the pull request:
This patch adds a preset that defaults the toolchain to use lld instead of gold on Linux.

Added presets are
  - `buildbot_linux,lld`
  - `buildbot_linux,lld,no_assertions`
  - `buildbot_linux,lld,no_test`
  - `buildbot_linux,lld,no_assertions,no_test`

etcwilde committed:
[build-script] Add buildbot_linux,lld preset

This patch adds a preset that defaults the toolchain to use lld instead
of gold on Linux.

Added presets are
  - `buildbot_linux,lld`
  - `buildbot_linux,lld,no_assertions`
  - `buildbot_linux,lld,no_test`
  - `buildbot_linux,lld,no_assertions,no_test`

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
preset=buildbot_linux,lld
@swift-ci please test with preset Linux Platform

etcwilde committed:
[build-script] Use use-linker flag

Switching presets to use the `use-linker` flag consistently instead of
setting the CMake variables directly.

etcwilde commented:
@swift-ci please test

etcwilde commented:
preset=buildbot_linux,lld
@swift-ci please test with preset Linux Platform

xavgru12 commented:
Why is the preset called build bot? Can this preset be used for local builds as well?

etcwilde commented:
> Why is the preset called build bot? Can this preset be used for local builds as well?

Staying consistent with what was there and to indicate that it's for the CI buildbots. Yes, you can use it for local builds if you want to replicate the configuration that CI is using to produce toolchains, though it might take longer than needed for normal development since it builds the entire toolchain instead of just pieces of it. I usually use a more stripped down `build-script --release --skip-test-cmark --skip-build-benchmarks` for building just the compiler, for instance.

xavgru12 commented:
Talking about more stripped down builds. What if I just want to build in debug mode: lib/sema (which is part of swift host tools afaik)? I thought —debug-swift would do it but it regresses. Can I build in debug single libraries like lib/sema without having to fiddle in CMakeLists.txt?

etcwilde commented:
Linux Preset Testing Failure:
```
ld.lld: error: unknown --compress-debug-sections value: zlib-gnu
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

etcwilde committed:
Update Backtracing/ElfReader.swift for lld

lld doesn't support zlib-gnu debug info compression. Gate that part of
the test on the ability to compile with that compression format.

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test
